### ISSUE TITLE
About the definition of reactor::init_new_scheduling_group_key ...

### ISSUE THREAD

**Issue by chinese20201103** [scylladb/seastar#903]
https://github.com/scylladb/scylla/issues/7746
https://github.com/scylladb/scylla/issues/8451
Scylla Server can start in dpdk mode after I modified the definition of reactor::init_new_scheduling_group_key!
```c++
future<>
reactor::init_new_scheduling_group_key(scheduling_group_key key, scheduling_group_key_config cfg) {
    auto& sg_data = _scheduling_group_specific_data;
    sg_data.scheduling_group_key_configs.resize(std::max<size_t>(sg_data.scheduling_group_key_configs.size(), key.id() + 1));
    sg_data.scheduling_group_key_configs[key.id()] = cfg;
/*	printf("_task_queues.size() %lu\n", _task_queues.size()); //wqf
    return parallel_for_each(_task_queues, [this, cfg, key] (std::unique_ptr<task_queue>& tq) {
        if (tq) {
			printf("tq->_id %d\n", tq->_id); //wqf
            scheduling_group sg = scheduling_group(tq->_id);
            return with_scheduling_group(sg, [this, key, sg] () {
                allocate_scheduling_group_specific_data(sg, key);
            });
        }
		else {
			printf("tq is null\n"); //wqf
		}
        return make_ready_future();
    });*/
	//wqf
	const auto prev_sg = current_scheduling_group();
	for (int i = 0; i < _task_queues.size(); i++) {
		auto* tq = _task_queues[i].get();
		if (tq) {
			scheduling_group sg = scheduling_group(tq->_id);
			*internal::current_scheduling_group_ptr() = sg;
			allocate_scheduling_group_specific_data(sg, key);
		}
	}
	*internal::current_scheduling_group_ptr() = prev_sg;
	return make_ready_future<>();
}
```

**Comment by nyh** [comment]
Thanks @chinese20201103 nice detective work!

Both the original code and your new code, at first view, appear correct. However, the main difference between them is that the original code does some of the work in the background. This is why I suspect maybe the bug has something to do with `reactor::init_new_scheduling_group_key()` not being reentrant - i.e., breaks if called twice in parallel:

 If the user code tries to add two different keys, it can lead to `init_new_scheduling_group_key()` being called for one key, and before it finishes its background work, a second `init_new_scheduling_group_key()` is called. But the second call may **resize the array**, which leads to moving the data, and the first call - still holding pointers to the elements of `_task_queue` in the various asynchronous functions it is running in parallel - will to look at a dead area.

If my guess is correct, wrapping the init_new_scheduling_group_key code in a semaphore that ensures only one copy can execute concurrently will work around this bug. But there has to be a better way :-)

Your approach is cool, just requires "internal" knowledge of how scheduling groups are remembered (instead of using the with_scheduling_group). Why is it critical in the first place that the `init_new_scheduling_group_key()` function runs in the scheduling group sg? Both the original code and your new code make an effort to ensure this, but I don't understand why.

The code your fixing, of per-scheduling-group data, was introduced by @eliransin in commit 1f68be436 - Eliran, what do you think?

**Comment by nyh** [comment]
A correction on my guess above:
Scylla only calls `scheduling_group_key_create()` once, so the problem can't be concurrent calls to this function. ..
My next guess is that we have one call to `scheduling_group_key_create()` concurrent with calls which create new scheduling groups (not keys). But I still couldn't figure out where this can break. It appears that `_task_queues` is a fixed-size vector, and shouldn't (?) move around when more task queues are activated.

**Comment by kostja** [comment]
Duplicate of https://github.com/scylladb/scylladb/issues/7746

**Comment by nyh** [comment]
@kostja, you marked this issue as a duplicate of an issue in a different repository....
